File: languages/java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/envelope/EnvelopeKeyRecordTest.java
package com.godaddy.asherah.appencryption.envelope;

import static org.junit.jupiter.api.Assertions.*;

import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.Base64;
import java.util.Optional;

import org.json.JSONObject;
import org.junit.jupiter.api.Test;

import com.godaddy.asherah.appencryption.utils.Json;

class EnvelopeKeyRecordTest {
  private final Instant created = Instant.now().truncatedTo(ChronoUnit.SECONDS);
  private final String parentKey = "key1";
  private final Instant parentCreated = created.minus(1, ChronoUnit.SECONDS);
  private final KeyMeta parentKeyMeta = new KeyMeta(parentKey, parentCreated);
  private final byte[] encryptedKey = new byte[] {0, 1, 2, 3, 4};
  private final boolean revoked = true;

  @Test
  void testConstructorRegularWithoutRevokedShouldBeOptionalEmpty() {
    EnvelopeKeyRecord record = new EnvelopeKeyRecord(created, parentKeyMeta, encryptedKey);
    assertNotNull(record);
    assertEquals(created, record.getCreated());
    assertEquals(Optional.of(parentKeyMeta), record.getParentKeyMeta());
    assertArrayEquals(encryptedKey, record.getEncryptedKey());
    assertEquals(Optional.empty(), record.isRevoked());
  }

  @Test
  void testConstructorRegularWithParentKeyMetaAndRevoked() {
    EnvelopeKeyRecord record = new EnvelopeKeyRecord(created, parentKeyMeta, encryptedKey, revoked);
    assertNotNull(record);
    assertEquals(created, record.getCreated());
    assertEquals(Optional.of(parentKeyMeta), record.getParentKeyMeta());
    assertArrayEquals(encryptedKey, record.getEncryptedKey());
    assertEquals(Optional.of(revoked), record.isRevoked());
  }

  @Test
  void testConstructorRegularWithNullParentKeyMetaAndNullRevokedShouldBeOptionalEmpty() {
    EnvelopeKeyRecord record = new EnvelopeKeyRecord(created, null, encryptedKey, null);
    assertNotNull(record);
    assertEquals(created, record.getCreated());
    assertEquals(Optional.empty(), record.getParentKeyMeta());
    assertArrayEquals(encryptedKey, record.getEncryptedKey());
    assertEquals(Optional.empty(), record.isRevoked());
  }

  @Test
  void testConstructorJsonWithParentKeyMetaAndRevoked() {
    // yes, i know this is exposing bad practice. meh
    Json parentKeyMetaJson = new Json();
    parentKeyMetaJson.put("KeyId", parentKey);
    parentKeyMetaJson.put("Created", parentCreated);

    Json envelopeKeyRecordJson = new Json();
    envelopeKeyRecordJson.put("Created", created);
    envelopeKeyRecordJson.put("ParentKeyMeta", parentKeyMetaJson);
    envelopeKeyRecordJson.put("Key", encryptedKey);
    envelopeKeyRecordJson.put("Revoked", revoked);

    EnvelopeKeyRecord record = new EnvelopeKeyRecord(envelopeKeyRecordJson);
    assertNotNull(record);
    assertEquals(created, record.getCreated());
    assertEquals(Optional.of(parentKeyMeta), record.getParentKeyMeta());
    assertArrayEquals(encryptedKey, record.getEncryptedKey());
    assertEquals(Optional.of(revoked), record.isRevoked());
  }

  @Test
  void testConstructorJsonWithNullParentKeyMetaAndNullRevokedShouldBeOptionalEmptyAndFalse() {
    Json envelopeKeyRecordJson = new Json();
    envelopeKeyRecordJson.put("Created", created);
    envelopeKeyRecordJson.put("Key", encryptedKey);

    EnvelopeKeyRecord record = new EnvelopeKeyRecord(envelopeKeyRecordJson);
    assertNotNull(record);
    assertEquals(created, record.getCreated());
    assertEquals(Optional.empty(), record.getParentKeyMeta());
    assertArrayEquals(encryptedKey, record.getEncryptedKey());
    assertEquals(Optional.of(false), record.isRevoked());
  }

  @Test
  void testToJsonWithParentKeyMetaAndRevoked() {
    EnvelopeKeyRecord envelopeKeyRecord = new EnvelopeKeyRecord(created, parentKeyMeta, encryptedKey, revoked);

    JSONObject recordJson = envelopeKeyRecord.toJson();
    assertEquals(created.getEpochSecond(), recordJson.get("Created"));
    // yes, i know this is exposing bad practice. meh
    assertEquals(parentCreated.getEpochSecond(), recordJson.getJSONObject("ParentKeyMeta").get("Created"));
    assertEquals(parentKey, recordJson.getJSONObject("ParentKeyMeta").getString("KeyId"));
    assertArrayEquals(encryptedKey, Base64.getDecoder().decode(recordJson.getString("Key")));
    assertEquals(revoked, recordJson.get("Revoked"));
  }

  @Test
  void testToJsonWithNullParentKeyMetaAndNullRevokedShouldBeNull() {
    EnvelopeKeyRecord envelopeKeyRecord = new EnvelopeKeyRecord(created, null, encryptedKey);

    JSONObject recordJson = envelopeKeyRecord.toJson();
    assertEquals(created.getEpochSecond(), recordJson.get("Created"));
    assertNull(recordJson.optJSONObject("ParentKeyMeta"));
    assertArrayEquals(encryptedKey, Base64.getDecoder().decode(recordJson.getString("Key")));
    assertNull(recordJson.opt("Revoked"));
  }

}

File: csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Envelope/EnvelopeKeyRecordTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Envelope;
using GoDaddy.Asherah.Crypto.ExtensionMethods;
using LanguageExt;
using Newtonsoft.Json.Linq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Envelope
{
    [Collection("Logger Fixture collection")]
    public class EnvelopeKeyRecordTest
    {
        private const string ParentKey = "key1";
        private const bool Revoked = true;
        private readonly DateTimeOffset created = DateTimeOffset.UtcNow.Truncate(TimeSpan.FromMinutes(1));
        private readonly byte[] encryptedKey = { 0, 1, 2, 3, 4 };
        private readonly DateTimeOffset parentCreated;
        private readonly KeyMeta parentKeyMeta;

        public EnvelopeKeyRecordTest()
        {
            // Have to init these in constructor
            parentCreated = created.AddSeconds(-1);
            parentKeyMeta = new KeyMeta(ParentKey, parentCreated);
        }

        [Fact]
        private void TestConstructorRegularWithoutRevokedShouldBeOptionalEmpty()
        {
            EnvelopeKeyRecord record = new EnvelopeKeyRecord(created, parentKeyMeta, encryptedKey);
            Assert.NotNull(record);
            Assert.Equal(created, record.Created);
            Assert.Equal(Option<KeyMeta>.Some(parentKeyMeta), record.ParentKeyMeta);
            Assert.Equal(encryptedKey, record.EncryptedKey);
            Assert.Equal(Option<bool>.None, record.Revoked);
        }

        [Fact]
        private void TestConstructorRegularWithParentKeyMetaAndRevoked()
        {
            EnvelopeKeyRecord record = new EnvelopeKeyRecord(created, parentKeyMeta, encryptedKey, Revoked);
            Assert.NotNull(record);
            Assert.Equal(created, record.Created);
            Assert.Equal(Option<KeyMeta>.Some(parentKeyMeta), record.ParentKeyMeta);
            Assert.Equal(encryptedKey, record.EncryptedKey);
            Assert.Equal(Option<bool>.Some(Revoked), record.Revoked);
        }

        [Fact]
        private void TestConstructorRegularWithNullParentKeyMetaAndNullRevokedShouldBeOptionalEmpty()
        {
            EnvelopeKeyRecord record = new EnvelopeKeyRecord(created, null, encryptedKey, null);
            Assert.NotNull(record);
            Assert.Equal(created, record.Created);
            Assert.Equal(Option<KeyMeta>.None, record.ParentKeyMeta);
            Assert.Equal(encryptedKey, record.EncryptedKey);
            Assert.Equal(Option<bool>.None, record.Revoked);
        }

        [Fact]
        private void TestConstructorJsonWithParentKeyMetaAndRevoked()
        {
            Asherah.AppEncryption.Util.Json parentKeyMetaJson = new Asherah.AppEncryption.Util.Json();
            parentKeyMetaJson.Put("KeyId", ParentKey);
            parentKeyMetaJson.Put("Created", parentCreated);

            Asherah.AppEncryption.Util.Json envelopeKeyRecordJson = new Asherah.AppEncryption.Util.Json();
            envelopeKeyRecordJson.Put("Created", created);
            envelopeKeyRecordJson.Put("ParentKeyMeta", parentKeyMetaJson);
            envelopeKeyRecordJson.Put("Key", encryptedKey);
            envelopeKeyRecordJson.Put("Revoked", Revoked);

            EnvelopeKeyRecord record = new EnvelopeKeyRecord(envelopeKeyRecordJson);
            Assert.NotNull(record);
            Assert.Equal(created, record.Created);
            Assert.Equal(Option<KeyMeta>.Some(parentKeyMeta), record.ParentKeyMeta);
            Assert.Equal(encryptedKey, record.EncryptedKey);
            Assert.Equal(Option<bool>.Some(Revoked), record.Revoked);
        }

        [Fact]
        private void TestConstructorJsonWithNullParentKeyMetaAndNullRevokedShouldBeOptionalEmptyAndFalse()
        {
            Asherah.AppEncryption.Util.Json envelopeKeyRecordJson = new Asherah.AppEncryption.Util.Json();
            envelopeKeyRecordJson.Put("Created", created);
            envelopeKeyRecordJson.Put("Key", encryptedKey);

            EnvelopeKeyRecord record = new EnvelopeKeyRecord(envelopeKeyRecordJson);
            Assert.NotNull(record);
            Assert.Equal(created, record.Created);
            Assert.Equal(Option<KeyMeta>.None, record.ParentKeyMeta);
            Assert.Equal(encryptedKey, record.EncryptedKey);
            Assert.Equal(Option<bool>.None, record.Revoked);
        }

        [Fact]
        private void TestToJsonWithParentKeyMetaAndRevoked()
        {
            EnvelopeKeyRecord envelopeKeyRecord = new EnvelopeKeyRecord(created, parentKeyMeta, encryptedKey, Revoked);

            JObject recordJson = envelopeKeyRecord.ToJson();
            Assert.Equal(created.ToUnixTimeSeconds(), recordJson.GetValue("Created").ToObject<long>());
            Assert.Equal(parentCreated.ToUnixTimeSeconds(), recordJson["ParentKeyMeta"]["Created"].ToObject<long>());
            Assert.Equal(ParentKey, recordJson["ParentKeyMeta"]["KeyId"].ToObject<string>());
            Assert.Equal(encryptedKey, Convert.FromBase64String(recordJson["Key"].ToString()));
            Assert.Equal(Revoked, recordJson["Revoked"].ToObject<bool>());
        }

        [Fact]
        private void TestToJsonWithNullParentKeyMetaAndNullRevokedShouldBeNull()
        {
            EnvelopeKeyRecord envelopeKeyRecord = new EnvelopeKeyRecord(created, null, encryptedKey);

            JObject recordJson = envelopeKeyRecord.ToJson();
            Assert.Equal(created.ToUnixTimeSeconds(), recordJson.GetValue("Created").ToObject<long>());
            Assert.Null(recordJson["ParentKeyMeta"]);
            Assert.Equal(encryptedKey, Convert.FromBase64String(recordJson["Key"].ToString()));
            Assert.Null(recordJson["Revoked"]);
        }
    }
}
